enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/ChunkedReplicatedContent.java

```diff
@@ -0,0 +1,132 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * Neo4j object code can be licensed independently from the source
+ * under separate terms from the AGPL. Inquiries can be directed to:
+ * licensing@neo4j.com
+ *
+ * More information is also available at:
+ * https://neo4j.com/licensing/
+ */
+package org.neo4j.causalclustering.messaging.marshalling;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.stream.ChunkedInput;
+
+import java.io.IOException;
+
+import org.neo4j.storageengine.api.WritableChannel;
+
+public class ChunkedReplicatedContent implements Marshal, ChunkedInput<ByteBuf>
+{
+
+    private static final int DEFAULT_CHUNK_SIZE = 8192;
+    private final byte contentType;
+    private final ByteBufAwareMarshal byteBufAwareMarshal;
+    private final int chunkSize;
+    private boolean lastByteWasWritten;
+    private int progress;
+
+    public ChunkedReplicatedContent( byte contentType, ByteBufAwareMarshal byteBufAwareMarshal, int chunkSize )
+    {
+        this.byteBufAwareMarshal = byteBufAwareMarshal;
+        this.chunkSize = chunkSize;
+        if ( chunkSize < 7 )
+        {
+            throw new IllegalArgumentException( "Chunk size must be at least 4 bytes" );
+        }
+        this.contentType = contentType;
+    }
+
+    public ChunkedReplicatedContent( byte contentType, ByteBufAwareMarshal byteBufAwareMarshal )
+    {
+        this( contentType, byteBufAwareMarshal, DEFAULT_CHUNK_SIZE );
+    }
+
+    @Override
+    public void marshal( WritableChannel channel ) throws IOException
+    {
+        channel.put( contentType );
+        byteBufAwareMarshal.marshal( channel );
+    }
+
+    @Override
+    public boolean isEndOfInput()
+    {
+        return lastByteWasWritten;
+    }
+
+    @Override
+    public void close()
+    {
+        // do nothing
+    }
+
+    @Override
+    public ByteBuf readChunk( ChannelHandlerContext ctx ) throws IOException
+    {
+        return readChunk( ctx.alloc() );
+    }
+
+    @Override
+    public ByteBuf readChunk( ByteBufAllocator allocator ) throws IOException
+    {
+        boolean endOfInput = isEndOfInput();
+        if ( endOfInput )
+        {
+            return null;
+        }
+        ByteBuf buffer = allocator.buffer( chunkSize );
+        try
+        {
+            buffer.writeBoolean( endOfInput );
+            if ( progress() == 0 )
+            {
+                // extra metadata on first chunk
+                buffer.writeByte( contentType );
+                buffer.writeInt( byteBufAwareMarshal.length() );
+            }
+            if ( !byteBufAwareMarshal.encode( buffer ) )
+            {
+                lastByteWasWritten = true;
+            }
+            if ( isEndOfInput() != endOfInput )
```

Comment:
This is really hard to read because of the overloaded usage with isEndOfInput() vs endOfInput vs lastByteWasWritten. Feels like it could be an else-clause, but I am not even certain what the logic is trying to achieve.

Reply:
So logic is. If, after writing to buffer, isEndOfInput has changed then we overwrite that value. Otherwise we leave it. But since all of the logic around this is handled in the same method I can agree that it could be simplified. Or just put the boolean last instead.